hutch_python/namespace.py

```diff
@@ -0,0 +1,108 @@
+from inspect import isfunction
+import logging
+
+from .utils import (IterableNamespace, find_class, strip_prefix,
+                    extract_objs)
+
+logger = logging.getLogger(__name__)
+
+
+def class_namespace(cls, scope=None):
+    """
+    Create a namespace that contains objects of a specific type.
+
+    Parameters
+    ----------
+    cls: type or str
+
+    scope: module, namespace, or list of these
+        Every object attached to the given modules will be considered for the
+        class_namespace. If scope is omitted, we'll check all objects loaded by
+        hutch_python and everything in the caller's global frame.
+
+    Returns
+    -------
+    namespace: IterableNamespace
+    """
+    logger.debug('Create class_namespace cls=%s, scope=%s', cls, scope)
+    class_space = IterableNamespace()
+    scope_objs = extract_objs(scope=scope, stack_offset=1)
+
+    if isinstance(cls, str):
+        if cls != 'function':
+            try:
+                cls = find_class(cls)
+            except Exception as exc:
+                err = 'Type {} could not be loaded'
+                logger.error(err.format(cls))
+                logger.debug(exc, exc_info=True)
+                return class_space
+
+    for name, obj in scope_objs.items():
+        include = False
+        if cls == 'function':
+            if isfunction(obj):
+                include = True
+        elif isinstance(obj, cls):
+            include = True
+        if include:
+            setattr(class_space, name, obj)
+            logger.debug('Include %s in cls=%s namespace', name, cls)
+
+    return class_space
+
+
+def metadata_namespace(md, scope=None):
```

Comment:
This is so much clearer. We can even have scientists write these live or in the `beamline.py` file